File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteProjectManager.java
package com.krystianwsul.checkme.firebase.records;

import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.util.Log;

import com.annimon.stream.Stream;
import com.google.firebase.database.DataSnapshot;
import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.firebase.DatabaseWrapper;
import com.krystianwsul.checkme.firebase.json.JsonWrapper;

import junit.framework.Assert;

import java.util.HashMap;
import java.util.Map;

public class RemoteProjectManager {
    private boolean mSaved = false;

    @NonNull
    public final Map<String, RemoteProjectRecord> mRemoteProjectRecords = new HashMap<>();

    public RemoteProjectManager(@NonNull DomainFactory domainFactory, @NonNull Iterable<DataSnapshot> children) {
        for (DataSnapshot child : children) {
            Assert.assertTrue(child != null);

            String key = child.getKey();
            Assert.assertTrue(!TextUtils.isEmpty(key));

            JsonWrapper jsonWrapper = child.getValue(JsonWrapper.class);
            Assert.assertTrue(jsonWrapper != null);

            RemoteProjectRecord remoteProjectRecord = new RemoteProjectRecord(domainFactory, key, jsonWrapper);

            mRemoteProjectRecords.put(key, remoteProjectRecord);
        }
    }

    public void save() {
        Map<String, Object> values = new HashMap<>();

        Stream.of(mRemoteProjectRecords.values())
                .forEach(remoteRecord -> remoteRecord.getValues(values));

        Log.e("asdf", "RemoteProjectManager.save values: " + values);

        if (!values.isEmpty()) {
            mSaved = true;
            DatabaseWrapper.updateRecords(values);
        }
    }

    public boolean isSaved() {
        return mSaved;
    }

    @NonNull
    public RemoteProjectRecord newRemoteProjectRecord(@NonNull DomainFactory domainFactory, @NonNull JsonWrapper jsonWrapper) {
        RemoteProjectRecord remoteProjectRecord = new RemoteProjectRecord(domainFactory, jsonWrapper);
        Assert.assertTrue(!mRemoteProjectRecords.containsKey(remoteProjectRecord.getId()));

        mRemoteProjectRecords.put(remoteProjectRecord.getId(), remoteProjectRecord);
        return remoteProjectRecord;
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteProjectManager.kt
package com.krystianwsul.checkme.firebase.records

import android.util.Log
import com.google.firebase.database.DataSnapshot
import com.krystianwsul.checkme.domainmodel.DomainFactory
import com.krystianwsul.checkme.firebase.DatabaseWrapper
import com.krystianwsul.checkme.firebase.json.JsonWrapper
import java.util.*

class RemoteProjectManager(private val domainFactory: DomainFactory, children: Iterable<DataSnapshot>) {

    private fun DataSnapshot.toRecord() = RemoteProjectRecord(domainFactory, key!!, getValue(JsonWrapper::class.java)!!)

    var isSaved = false

    val remoteProjectRecords = children.associate { child -> child.key!! to child.toRecord() }.toMutableMap()

    fun addChild(dataSnapshot: DataSnapshot): RemoteProjectRecord {
        val key = dataSnapshot.key!!
        check(!remoteProjectRecords.containsKey(key))

        return dataSnapshot.toRecord().also {
            remoteProjectRecords[key] = it
        }
    }

    fun changeChild(dataSnapshot: DataSnapshot): RemoteProjectRecord {
        val key = dataSnapshot.key!!
        check(remoteProjectRecords.containsKey(key))

        return dataSnapshot.toRecord().also {
            remoteProjectRecords[key] = it
        }
    }

    fun removeChild(dataSnapshot: DataSnapshot) = dataSnapshot.key!!.also {
        check(remoteProjectRecords.remove(it) != null)
    }

    fun save(): Boolean {
        val values = HashMap<String, Any?>()

        remoteProjectRecords.values.forEach { it.getValues(values) }

        Log.e("asdf", "RemoteProjectManager.save values: $values")

        if (!values.isEmpty()) {
            check(!isSaved)

            Log.e("asdf", "saving task records:\n$values")

            isSaved = true
            DatabaseWrapper.updateRecords(values)
        }

        return isSaved
    }

    fun newRemoteProjectRecord(domainFactory: DomainFactory, jsonWrapper: JsonWrapper) = RemoteProjectRecord(domainFactory, jsonWrapper).also {
        check(!remoteProjectRecords.containsKey(it.id))

        remoteProjectRecords[it.id] = it
    }
}
